File: packages/engine-freecut/src/executor/assetResolver.ts
// packages/engine-freecut/src/executor/assetResolver.ts
import fs from "fs/promises";
import path from "path";

export interface AssetEntry {
  mediaId: string;            // canonical id passed to actions
  semanticName?: string;      // human-readable, for prompt context
  filePath: string;           // ABSOLUTE path on disk
  kind: "video" | "audio" | "image";
  durationSec?: number;
}

export class AssetResolver {
  private byId = new Map<string, AssetEntry>();

  constructor(entries: AssetEntry[] = []) {
    for (const e of entries) this.byId.set(e.mediaId, e);
  }

  register(entry: AssetEntry) {
    this.byId.set(entry.mediaId, entry);
  }

  resolve(mediaId: string): AssetEntry | undefined {
    return this.byId.get(mediaId);
  }

  async assertAllExist(mediaIds: string[]): Promise<{
    resolved: Record<string, string>;
    unresolved: string[];
  }> {
    const resolved: Record<string, string> = {};
    const unresolved: string[] = [];

    for (const id of mediaIds) {
      const entry = this.byId.get(id);
      if (!entry) {
        unresolved.push(id);
        continue;
      }
      try {
        await fs.access(entry.filePath);
        resolved[id] = entry.filePath;
      } catch {
        unresolved.push(`${id} (missing file: ${entry.filePath})`);
      }
    }
    return { resolved, unresolved };
  }

  /** Used by the planner to give Gemini ONLY real asset ids */
  toPromptContext(): string {
    const lines = ["AVAILABLE ASSETS (use these exact mediaId values):"];
    for (const e of this.byId.values()) {
      lines.push(
        `- mediaId="${e.mediaId}" kind=${e.kind}` +
          (e.semanticName ? ` description="${e.semanticName}"` : "") +
          (e.durationSec ? ` duration=${e.durationSec.toFixed(2)}s` : "")
      );
    }
    return lines.join("\n");
  }
}

File: monet/engines/freecut/executor/asset_resolver.py
# monet/engines/freecut/executor/asset_resolver.py
from __future__ import annotations
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, List, Literal, Optional, Tuple


@dataclass
class AssetEntry:
    mediaId: str
    filePath: str
    kind: Literal["video", "audio", "image"]
    semanticName: Optional[str] = None
    durationSec: Optional[float] = None


class AssetResolver:
    def __init__(self, entries: Optional[List[AssetEntry]] = None):
        self._by_id: Dict[str, AssetEntry] = {}
        for e in entries or []:
            self.register(e)

    def register(self, entry: AssetEntry) -> None:
        self._by_id[entry.mediaId] = entry

    def resolve(self, media_id: str) -> Optional[AssetEntry]:
        return self._by_id.get(media_id)

    def assert_all_exist(self, media_ids: List[str]) -> Tuple[Dict[str, str], List[str]]:
        resolved: Dict[str, str] = {}
        unresolved: List[str] = []
        for mid in media_ids:
            e = self._by_id.get(mid)
            if not e:
                unresolved.append(mid)
                continue
            if not Path(e.filePath).exists():
                unresolved.append(f"{mid} (missing file: {e.filePath})")
                continue
            resolved[mid] = e.filePath
        return resolved, unresolved

    def to_prompt_context(self) -> str:
        lines = ["AVAILABLE ASSETS (use these exact mediaId values):"]
        for e in self._by_id.values():
            parts = [f'mediaId="{e.mediaId}"', f"kind={e.kind}"]
            if e.semanticName:
                parts.append(f'description="{e.semanticName}"')
            if e.durationSec is not None:
                parts.append(f"duration={e.durationSec:.2f}s")
            lines.append(f"- {' '.join(parts)}")
        return "\n".join(lines)
